[Drake]Vending step6

DrakeYang opened the pull request:
피드백 부탁드립니다!

DrakeYang committed:
스텝 5 정리 및 스텝6 추가

DrakeYang committed:
구매버튼 6개 추가. 태그설정.
기존 사진 150x100 으로 사이즈 조절. 전체적인 위치 조절

DrakeYang committed:
음료사진 radius 15 로 변경

DrakeYang committed:
음료사진 borderWidth 5 로 설정

DrakeYang committed:
음료주문 함수에서 중복된 부분 삭제

DrakeYang committed:
음료주문 함수 추가 : 음료 태그를 받아서 진행

DrakeYang committed:
음료 추가 로직 에러수정

DrakeYang committed:
음료창고 주석 수정

DrakeYang committed:
음료 주문 버튼 추가

DrakeYang committed:
음료 주문시에도 음료재고변동 노티를 보내도록 추가

DrakeYang committed:
잔액변동 노티를 추가

DrakeYang committed:
주문된 음료를 담는 객체 추가 : OrderedDrinkSlot

DrakeYang committed:
인코딩,디코딩,음료추가,태그리턴 함수 생성

DrakeYang committed:
프로젝트 파일 수정됨 : xcode 에서 파일위치 이동

DrakeYang committed:
주문된 음료가 없을시 빈배열 리턴 : 이전엔 에러던짐

DrakeYang committed:
음료를 배열로 리턴하는 함수 추가

DrakeYang committed:
음료슬롯을 받아서 음료를 추가하는 함수 추가

DrakeYang committed:
주문된음료 객체 타입 변화로 인코딩 키 변경

DrakeYang committed:
주문된음료 객체 변경 : 음료창고 > 주문된음료.
음료 주문 함수 주문된음료 객체로 수정

DrakeYang committed:
함수 첫글자 소문자로 수정 : allOderedDrinksTag

DrakeYang committed:
주문된음료 음료태그 리턴 함수와 연결되는 함수 추가

DrakeYang committed:
금액비교 함수에 재고체크도 추가

DrakeYang committed:
음료주문 노티 추가

DrakeYang committed:
음료 주문시 주문된 음료사진 추가함수 추가.
음료주문 관련 노티 추가

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -72,8 +76,8 @@ class VendingMachine : NSObject, vendinMachineMenu, NSCoding  {
-    func getAllOrderdDrink()->String{
-        return orderedDrinks.getTotalDrinkDetail().getAllDrinkDetails()
+    func getAllOrderdDrink()->[Int]{
+        return orderedDrinks.allOderedDrinksTag()
```
왜 이런거부터 보이는지 ... allOrder- 겠죠?

DrakeYang replied:
앗 6시 전까지 보내려고 너무 서둘렀나봅니다...!

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -121,11 +129,31 @@ extension VendingMachine {
+        // 음료재고변동 노티를 보낸다
+        NotificationCenter.default.post(name: .drinkCountChanged, object: nil)
+        // 음료재고변동 노티를 보낸다
+        NotificationCenter.default.post(name: .orderDrink, object: nil)
```
노티를 여기서 처리하는 것을 두 가지 측면에서 고민해봐야 합니다.
- 바로 위에 `try buyDrink()` 에서 실패하면 어떻게 되죠? drinkCount 가 변경이 되나요? 
- orderDrink 라는 이벤트는 음료개수가 바뀐거와 같은건가요 별개인가요? 
- 리턴하는 result가 있으면 노티가 필요없나요? 그래도 필요한가요?

DrakeYang replied:
노티를 재고추가, 재고추출(재고 pop), 주문한음료 추가 로 나누었고,
해당 변경이 일하는 직후에 노티를 보내도록 수정했습니다.
원래는 실패하면 재고가 변경 되지만 표시는 바뀌지 않았습니다. 
수정 이후엔 재고가 변경되고 나서 함수고 에러가 나도 노티에서 에러가 나지 않는다면 표시는 변합니다.
리턴하는 결과가 있더라도 노티는 필요한것 같습니다. 노티를 리턴에 묶으면 리턴에서 에러가 나면 표시되는 값 변환이 안될것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -183,7 +211,12 @@ extension VendingMachine {
         _ = try addDrink(drink: newDrink)
-        // 음료 추가가 완료 되었다는 노티를 보낸다
+        // 음료재고변동 노티를 보낸다
         NotificationCenter.default.post(name: .drinkCountChanged, object: nil)
```
노티 `drinkCountChanged` 는 추가할 때도, 구매할 때도 사용하면 구분할 필요는 없나요?

DrakeYang replied:
저도 그렇게 생각합니다 ㅋㅋㅋ 노티를 사용하라고 하셔서 굳이넣긴 했는데 사실 중복되는 부분이죠!
삭제하겠습니다.

DrakeYang replied:
노티를 재고추가, 재고추출(재고 pop), 주문한음료 추가 로 나누었습니다.
중첩되는 부분은 삭제했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachineDataManager.swift:
```diff
@@ -10,7 +10,7 @@ import Foundation
 class VendingMachineDataManager {
     
-    private static let dataKey = "dataKey"
+    private static let dataKey = "dataKey2"
```
"dataKey2" 라는 건 어떤 의미죠? 1이 있나요?

DrakeYang replied:
주문된 음료 저장객체의 타입을 바꿨더니 에러가 발생하더라구요
그래서 키를 다르게 주었습니다.
초기화를 하고 키는 바꾸지 않는게 맞는건가요?

godrm replied:
이미 저장된 객체 때문에 그렇게했던건가 보군요. 
초기화 하는게 맞겠죠. 그렇다고 코드를 고치면 안되죠!
시뮬레이터에서 에러가 나는건 앱을 지우고 다시 설치하면 됩니다.

DrakeYang replied:
아 이런 방법이 !! 적용하겠습니다!

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -45,7 +47,22 @@ class ViewController: UIViewController {
+    /// 음료구매버튼 액션
+    @IBAction func buyDrink(_ sender: UIButton) {
+        do {
+            _ = try vendingMachine.buyDrink(drinkTag: sender.tag-20)
```
buyDrink()를 사용할 때 tag를 활용하고 있는데 자판기 로직이 화면 요소와 너무 연관된 느낌입니다.
tag 값과 매칭하는 것은 화면 순서 때문인거지 vendingMachine과 관련이 없습니다.
모델과 뷰 각각에 필요한 내용이죠. 이걸 매칭하는 걸 자판기로 넘기지 말고 별도로 생각해야 합니다.

DrakeYang replied:
맞습니다! 자판기가 외부 개념인 음료태그를 알 필요가 전혀 없더라구요.
자판기에서 음료타입을 받아서 작동하는 방식으로 변경했고, 음료태그를 음료타입으로 변환하는 함수를 뷰컨에 추가해서 돌아가도록 했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -128,7 +148,42 @@ class ViewController: UIViewController {
+            self.view.addSubview(cardImage)
+            // 음료 카운트 추가
+            drinkCount += 1
+        }
```
refreshOrderedDrink() 를 호출할 때마다 계속 addSubview()를 호출하고 있는 것 같습니다.
이러면 반복해서 계속 뷰가 추가되는 현상이 생깁니다. 처음에만 추가하고, 새로 갱신하는게 가장 좋습니다.
그렇지 않으면 일부 뷰들만 remove 하고 다시 추가하도록 관리가 필요합니다. 
이 부분을 확인해보세요. 어떻게 확인할 수 있을까요?

DrakeYang replied:
처음 전체를 부르는 부분과 추가하는 부분을 함수로 분리시켜보겠습니다.

DrakeYang replied:
앱 실행시 주문된음료 전체를 부르는 함수와 추가시 추가된 음료만 추가하는 함수로 나누었습니다.

godrm reviewed (commented):
오늘 설명한것처럼 뷰 -> 컨트롤러 > 모델로 데이터를 업데이트 하는 흐름과
모델 ----> 컨트롤러 ----> 뷰까지 화면을 업데이트 하는 흐름을 구분해서 단방향으로 진행되도록 설계해야 합니다.
다음 피드백을 확인해보세요.

DrakeYang committed:
오타수정 : allOrderedDrinksTag

DrakeYang committed:
중복되는 음료주문 노티 삭제

DrakeYang committed:
주문된 음료 사진추가 함수 기능분리 : 시작시 한번 돌리는 함수, 반복용 함수

DrakeYang committed:
노티관련 함수들 변경 : 재고변경시 > 음료추가시

DrakeYang committed:
주문된 음료 관련 노티 추가 및 수정.
음료주문 완료시 음료주문 완료 노티를 보냅 > 노티가 발생하면 재고갱신,주문된음료 갠신 함수가 작동.
주문된 음료를 모두 띄우는 함수는 시작시 1번만 실행. 이후에는 추가함수만 실행

DrakeYang committed:
음료재고변동과 주문한음료리스트 노티 분리.
음료재고 추가와 삭제 노티를 분리.
주문한음료 노티 분리. 노티 이름 수정 : orderDrink > afterAddOrderedDrinkList.
음료 재고 삭제 바로 뒤에 재고삭제 노티 추가, 주문한음료 추가 직후에 노티추가

DrakeYang committed:
자판기 주문함수 입련인자 변경 : 음료태그 > 음료타입.
자판기가 음료태그를 알필요가 없음

DrakeYang committed:
음료주문 함수 단계 추가.
음료태그를 받아서 음료타입으로 변환하는 함수 추가.
음료주문 시 태그를 타입으로 변환하는 단계 추가

DrakeYang commented:
피드백 부탁드립니다!
